File: Android/scenes/show/src/main/java/io/agora/scene/show/service/rtmsync/PKService.kt
package io.agora.scene.show.service.rtmsync

import androidx.annotation.IntDef
import io.agora.rtmsyncmanager.SyncManager
import io.agora.rtmsyncmanager.model.AUIRoomContext
import io.agora.rtmsyncmanager.utils.AUILogger
import io.agora.rtmsyncmanager.utils.GsonTools
import io.agora.rtmsyncmanager.utils.ObservableHelper
import java.util.UUID

class PKService(
    private val channelName: String,
    private val syncManager: SyncManager,
    private val roomPresenceService: RoomPresenceService,
    private val interactionService: InteractionService
) {
    private val tag = "PKService($channelName)"
    private val key = "pk"

    private val observerHelper = ObservableHelper<(PKInfo) -> Unit>()

    private val messageRetainer = MessageRetainer(syncManager.rtmManager, channelName, key).apply {
        subscribe { msg ->
            val pkInfo = GsonTools.toBeanSafely(
                msg.content,
                PKInfo::class.java
            )
            AUILogger.logger().d(tag, "onPKUpdated from message: $pkInfo")
            if (pkInfo != null) {
                observerHelper.notifyEventHandlers {
                    it.invoke(pkInfo)
                }
            }
        }
    }

    private val roomPresenceSubscriber = RoomPresenceSubscriber(
        onUpdate = { info: RoomPresenceInfo ->
            AUILogger.logger().d(tag, "onRoomPresenceUpdated: $info")

            // 被PK方 -> 发起PK方
            if (info.roomId != channelName) {
                val currInfo = roomPresenceService.getRoomPresenceInfo(channelName)
                if (info.status == RoomPresenceStatus.INTERACTING_PK
                    && currInfo?.status == RoomPresenceStatus.IDLE
                    && info.interactorId == AUIRoomContext.shared().currentUserInfo.userId
                ) {
                    AUILogger.logger().d(tag, "after acceptPK >> updateRoomPresenceInfo : status=${RoomPresenceStatus.INTERACTING_PK}, interactorId=${info.ownerId}, interactorName=${info.ownerName}")
                    roomPresenceService.updateRoomPresenceInfo(
                        channelName,
                        status = RoomPresenceStatus.INTERACTING_PK,
                        interactorId = info.ownerId,
                        interactorName = info.ownerName,
                    )
                }
            }

            // 发起PK方 -> 被PK方
            if (info.roomId != channelName) {
                val currInfo = roomPresenceService.getRoomPresenceInfo(channelName)
                if (
                    info.status == RoomPresenceStatus.INTERACTING_PK
                    && currInfo?.status == RoomPresenceStatus.INTERACTING_PK
                    && currInfo.interactorId == info.ownerId
                    && info.interactorId == currInfo.ownerId
                ) {
                    AUILogger.logger().d(tag, "after acceptPK >> startPKInteraction roomId=${info.roomId}, ownerId=${info.ownerId}, ownerName=${info.ownerName}")
                    interactionService.startPKInteraction(
                        info.roomId,
                        info.ownerId,
                        info.ownerName,
                    )
                    observerHelper.notifyEventHandlers {
                        it.invoke(
                            PKInfo(
                                userId = info.ownerId,
                                userName = info.ownerName,
                                roomId = info.roomId,
                                fromUserId = info.interactorId,
                                fromUserName = info.interactorName,
                                fromRoomId = info.roomId,
                                type = PKType.ACCEPT
                            )
                        )
                    }
                }
            }

            // 发起PK方
            if (info.roomId == channelName) {
                val pkRoomInfo = roomPresenceService.getRoomPresenceInfoByOwnerId(info.interactorId)
                if (
                    info.status == RoomPresenceStatus.INTERACTING_PK
                    && pkRoomInfo?.status == RoomPresenceStatus.INTERACTING_PK
                    && pkRoomInfo.interactorId == info.ownerId
                    && info.interactorId == pkRoomInfo.ownerId
                ) {
                    AUILogger.logger().d(tag, "after acceptPK >> startPKInteraction roomId=${pkRoomInfo.roomId}, ownerId=${pkRoomInfo.ownerId}, ownerName=${pkRoomInfo.ownerName}")
                    interactionService.startPKInteraction(
                        pkRoomInfo.roomId,
                        pkRoomInfo.ownerId,
                        pkRoomInfo.ownerName
                    )
                    observerHelper.notifyEventHandlers {
                        it.invoke(
                            PKInfo(
                                userId = pkRoomInfo.ownerId,
                                userName = pkRoomInfo.ownerName,
                                roomId = pkRoomInfo.roomId,
                                fromUserId = pkRoomInfo.interactorId,
                                fromUserName = pkRoomInfo.interactorName,
                                fromRoomId = pkRoomInfo.roomId,
                                type = PKType.ACCEPT
                            )
                        )
                    }
                }
            }

            // 被PK方/发起PK方 有一方关闭即停止PK
            if (info.roomId != channelName) {
                val currInfo = roomPresenceService.getRoomPresenceInfo(channelName)
                if (
                    info.status == RoomPresenceStatus.IDLE
                    && currInfo?.status == RoomPresenceStatus.INTERACTING_PK
                    && currInfo.interactorId == info.ownerId
                ) {
                    AUILogger.logger().d(tag, "after stopPK >> stopInteraction")
                    interactionService.stopInteraction(
                        success = {
                            AUILogger.logger().d(tag, "after stopPK >> stopInteraction success")
                        },
                        failure = {
                            AUILogger.logger().d(tag, "after stopPK >> stopInteraction error: $it")
                        }
                    )
                    observerHelper.notifyEventHandlers {
                        it.invoke(
                            PKInfo(
                                userId = currInfo.ownerId,
                                userName = currInfo.ownerName,
                                roomId = currInfo.roomId,
                                fromUserId = info.ownerId,
                                fromUserName = info.ownerName,
                                fromRoomId = info.roomId,
                                type = PKType.END
                            )
                        )
                    }
                }
            }

            // 主播已经和其他人PK
            if (info.roomId != channelName) {
                val currInfo = roomPresenceService.getRoomPresenceInfo(channelName)
                if (info.status == RoomPresenceStatus.INTERACTING_PK
                    && currInfo?.status == RoomPresenceStatus.INTERACTING_PK
                    && currInfo.interactorId == info.ownerId
                    && info.interactorId != AUIRoomContext.shared().currentUserInfo.userId
                ) {
                    AUILogger.logger().d(tag, "after startPK >> updateRoomPresenceInfo for pk owner has started interaction with others: status=${RoomPresenceStatus.IDLE}")
                    roomPresenceService.updateRoomPresenceInfo(
                        channelName,
                        status = RoomPresenceStatus.IDLE,
                    )
                    observerHelper.notifyEventHandlers {
                        it.invoke(
                            PKInfo(
                                userId = currInfo.ownerId,
                                userName = currInfo.ownerName,
                                roomId = currInfo.roomId,
                                fromUserId = currInfo.interactorId,
                                fromUserName = currInfo.interactorName,
                                fromRoomId = currInfo.roomId,
                                type = PKType.END
                            )
                        )
                    }
                }
            }
        },
        onDelete = {
            val interactionInfo = interactionService.getInteractionInfo()
            if (interactionInfo?.type == InteractionType.PK && it.ownerId == interactionInfo.userId) {
                interactionService.stopInteraction()
                roomPresenceService.updateRoomPresenceInfo(
                    channelName,
                    status = RoomPresenceStatus.IDLE,
                )
            }
        }
    )

    init {
        roomPresenceService.subscribe(roomPresenceSubscriber)
    }


    fun invitePK(
        roomId: String,
        success: (PKInfo) -> Unit,
        error: ((Throwable) -> Unit)? = null
    ) {
        val roomPresenceInfo = roomPresenceService.getRoomPresenceInfo(roomId)
        if (roomPresenceInfo == null) {
            AUILogger.logger().d(tag, "invitePK >> room presence info is null")
            error?.invoke(RuntimeException("room presence info is null"))
            return
        }
        if (roomPresenceInfo.status != RoomPresenceStatus.IDLE) {
            AUILogger.logger().d(tag, "invitePK >> room presence status is not idle")
            error?.invoke(RuntimeException("room presence status is not idle"))
            return
        }
        val currRoomPresenceInfo = roomPresenceService.getRoomPresenceInfo(channelName)
        if(currRoomPresenceInfo?.status != RoomPresenceStatus.IDLE){
            AUILogger.logger().d(tag, "invitePK >> current room presence status is not idle")
            error?.invoke(RuntimeException("current room presence status is not idle"))
            return
        }
        val pkInfo = PKInfo(
            userId = roomPresenceInfo.ownerId,
            userName = roomPresenceInfo.ownerName,
            roomId = roomPresenceInfo.roomId,
            fromUserId = AUIRoomContext.shared().currentUserInfo.userId,
            fromUserName = AUIRoomContext.shared().currentUserInfo.userName,
            fromRoomId = channelName,
            type = PKType.INVITING
        )
        AUILogger.logger().d(tag, "[${pkInfo.id}] invitePK >> sendMessage : $pkInfo")
        messageRetainer.sendMessage(
            GsonTools.beanToString(pkInfo) ?: "",
            roomPresenceInfo.ownerId,
            roomPresenceInfo.roomId,
            success = {
                AUILogger.logger().d(tag, "[${pkInfo.id}] invitePK >> sendMessage success")
                success.invoke(pkInfo)
            },
            error = {
                AUILogger.logger().d(tag, "[${pkInfo.id}] invitePK >> sendMessage error: $it")
                error?.invoke(it)
            }
        )
    }

    fun acceptPK(
        pkId: String,
        success: (() -> Unit)? = null,
        error: ((Throwable) -> Unit)? = null
    ) {
        val message = messageRetainer.getMessage {
            GsonTools.toBeanSafely(
                it.content,
                PKInfo::class.java
            )?.id == pkId
        }
        val pkInfo = GsonTools.toBeanSafely(
            message?.content,
            PKInfo::class.java
        )
        if (message == null || pkInfo == null) {
            AUILogger.logger().d(tag, "[$pkId] acceptPK >> pk info is null")
            error?.invoke(RuntimeException("pk info is null"))
            return
        }
        val roomPresenceInfo = roomPresenceService.getRoomPresenceInfo(pkInfo.fromRoomId)
        if (roomPresenceInfo?.status != RoomPresenceStatus.IDLE) {
            AUILogger.logger().d(tag, "[$pkId] acceptPK >> room presence status is not idle")
            error?.invoke(RuntimeException("room presence status is not idle"))
            return
        }
        // 因为不管对方在不在线，更新presence总是会成功，这里被一个点对点消息用于判断对方是否在线
        messageRetainer.sendMessage(
            GsonTools.beanToString(pkInfo.copy(type = PKType.ACCEPT)) ?: "",
            pkInfo.fromUserId,
            pkInfo.fromRoomId,
            success = {
                AUILogger.logger().d(tag, "[$pkId] acceptPK >> sendMessage success")
                AUILogger.logger().d(tag, "[$pkId] acceptPK >> updateRoomPresenceInfo : interactorId=${pkInfo.fromUserId}, interactorName=${pkInfo.fromUserName}")
                roomPresenceService.updateRoomPresenceInfo(
                    channelName,
                    status = RoomPresenceStatus.INTERACTING_PK,
                    interactorId = pkInfo.fromUserId,
                    interactorName = pkInfo.fromUserName,
                    success = {
                        AUILogger.logger().d(tag, "[$pkId] acceptPK >> updateRoomPresenceInfo success")
                        messageRetainer.removeMessage(message.id)
                        success?.invoke()
                    },
                    error = {
                        AUILogger.logger().d(tag, "[$pkId] acceptPK >> updateRoomPresenceInfo error: $it")
                        error?.invoke(RuntimeException(it))
                    }
                )
            },
            error = {
                AUILogger.logger().d(tag, "[$pkId] acceptPK >> sendMessage error: $it")
                error?.invoke(it)
            }
        )
    }

    fun rejectPK(
        pkId: String,
        success: (() -> Unit)? = null,
        error: ((Throwable) -> Unit)? = null
    ) {
        val message = messageRetainer.getMessage {
            GsonTools.toBeanSafely(
                it.content,
                PKInfo::class.java
            )?.id == pkId
        }
        var pkInfo = GsonTools.toBeanSafely(
            message?.content,
            PKInfo::class.java
        )
        if (message == null || pkInfo == null) {
            AUILogger.logger().d(tag, "[$pkId] rejectPK >> pk info is null")
            error?.invoke(RuntimeException("pk info is null"))
            return
        }

        pkInfo = pkInfo.copy(type = PKType.REJECT)
        AUILogger.logger().d(tag, "[$pkId] rejectPK >> sendMessage : $pkInfo")
        messageRetainer.sendMessage(
            GsonTools.beanToString(pkInfo) ?: "",
            pkInfo.fromUserId,
            pkInfo.fromRoomId,
            success = {
                AUILogger.logger().d(tag, "[$pkId] rejectPK >> sendMessage success")
                messageRetainer.removeMessage(message.id)
                success?.invoke()
            },
            error = {
                AUILogger.logger().d(tag, "[$pkId] rejectPK >> sendMessage error: $it")
                error?.invoke(it)
            }
        )
    }

    fun subscribe(onUpdate: (PKInfo) -> Unit) {
        observerHelper.subscribeEvent(onUpdate)
    }

    fun unSubscribe(onUpdate: (PKInfo) -> Unit) {
        observerHelper.unSubscribeEvent(onUpdate)
    }

    fun release() {
        AUILogger.logger().d(tag, "release")
        observerHelper.unSubscribeAll()
        roomPresenceService.unSubscribe(roomPresenceSubscriber)
        messageRetainer.release()
    }

}

@IntDef(PKType.INVITING, PKType.ACCEPT, PKType.REJECT, PKType.END)
@Target(AnnotationTarget.FIELD)
@Retention(AnnotationRetention.SOURCE)
annotation class PKType {
    companion object {
        const val INVITING = 0
        const val ACCEPT = 1
        const val REJECT = 2
        const val END = 3
    }
}

data class PKInfo(
    val id: String = UUID.randomUUID().toString(),
    val userId: String,
    var userName: String,
    val roomId: String,
    val fromUserId: String,
    val fromUserName: String,
    val fromRoomId: String,
    @PKType val type: Int = PKType.INVITING,
)

File: iOS/AgoraEntScenarios/Common/API/RTMSyncManager/RTMSyncManager/Classes/Extension/ApplyAndInvitation/PKService.swift
//
//  PKService.swift
//  RTMSyncManager
//
//  Created by wushengtao on 2024/6/7.
//

import Foundation

@objc public enum PKType: Int, Codable {
    case inviting = 0
    case accept = 1
    case reject = 2
    case end = 3
}

@objcMembers public class PKInfo: NSObject, Codable {
    public var id: String = UUID().uuidString
    public var userId: String = ""
    public var userName: String = ""
    public var roomId: String = ""
    public var fromUserId: String = ""
    public var fromUserName: String = ""
    public var fromRoomId: String = ""
    public var type: PKType = .inviting
    
    enum CodingKeys: String, CodingKey {
        case id, userId, userName, roomId, fromUserId, fromUserName, fromRoomId, type
    }
}

@objc public protocol PKServiceProtocol: NSObjectProtocol {
    func onPKInfoDidReceive(channelName: String, info: PKInfo)
}

private let key = "pk"

public class PKService: NSObject {
    private var channelName: String = ""
    private var syncManager: AUISyncManager
    private var roomPresenceService: RoomPresenceService
    private var interactionService: InteractionService
    
    private var respDelegates = NSHashTable<PKServiceProtocol>.weakObjects()
    
    private lazy var messageManager: InviteMessageManager = {
        let manager = InviteMessageManager(channelName: channelName, key: key, rtmManager: syncManager.rtmManager)
        return manager
    }()
    
    deinit {
        aui_info("deinit PKService[\(channelName)]", tag: "PKService")
        innerUnsubscribe()
    }
    
    required init(channelName: String,
                  syncManager: AUISyncManager,
                  presenceService: RoomPresenceService,
                  interactionService: InteractionService) {
        aui_info("init PKService[\(channelName)]", tag: "PKService")
        self.channelName = channelName
        self.syncManager = syncManager
        self.roomPresenceService = presenceService
        self.interactionService = interactionService
        super.init()
        innerSubscribe()
    }
    
    private func innerSubscribe() {
        messageManager.subscribe(delegate: self)
        roomPresenceService.subscribe(delegate: self)
    }
    
    private func innerUnsubscribe() {
        messageManager.unsubscribe(delegate: self)
        roomPresenceService.unsubscribe(delegate: self)
    }
    
    private func createPKInfo(_ user: RoomPresenceInfo, type: PKType) -> PKInfo {
        let info = PKInfo()
        info.userId = user.ownerId
        info.userName = user.ownerName
        info.roomId = user.roomId
        info.fromUserId = user.interactorId
        info.fromUserName = user.interactorName
        info.fromRoomId = user.roomId
        info.type = type
        
        return info
    }
}

//MARK: public
extension PKService {
    public func subscribe(delegate: PKServiceProtocol) {
        if respDelegates.contains(delegate) {
            return
        }
        
        respDelegates.add(delegate)
    }
    
    public func unsubscribe(delegate: PKServiceProtocol) {
        respDelegates.remove(delegate)
    }
    
    public func invitePK(roomId: String, completion: ((NSError?) -> ())?) {
        guard let roomPresenceInfo = roomPresenceService.getRoomPresenceInfo(roomId: roomId),
              roomPresenceInfo.status == .idle else {
            aui_info("invitePK roomId: \(roomId) fail: room presence status is not idle", tag: "PKService")
            completion?(NSError(domain: "room presence status is not idle", code: 0, userInfo: nil))
            return
        }
        
        guard let currRoomPresenceInfo = roomPresenceService.getRoomPresenceInfo(roomId: channelName),
              currRoomPresenceInfo.status == .idle else {
            aui_info("invitePK roomId: \(roomId) fail: current room presence status is not idle", tag: "PKService")
            completion?(NSError(domain: "current room presence status is not idle", code: 0, userInfo: nil))
            return
        }
        
        aui_info("invitePK roomId: \(roomId)", tag: "PKService")
        let info = PKInfo()
        info.userId = roomPresenceInfo.ownerId
        info.userName = roomPresenceInfo.ownerName
        info.roomId = roomPresenceInfo.roomId
        info.fromUserId = AUIRoomContext.shared.currentUserInfo.userId
        info.fromUserName = AUIRoomContext.shared.currentUserInfo.userName
        info.fromRoomId = channelName
        info.type = .inviting
        
        let content = encodeModelToJsonStr(info) ?? ""
        messageManager.sendMessage(content: content,
                                   userId: roomPresenceInfo.ownerId,
                                   channelName: roomPresenceInfo.roomId) { err in
            aui_info("invitePK roomId: \(roomId) completion: \(err?.localizedDescription ?? "success")", tag: "PKService")
            completion?(err)
        }
    }
    
    public func acceptPK(pkId: String, completion: ((NSError?) -> ())?) {
        guard let message = messageManager.getMessage({ $0.content.contains(pkId) }),
              let pkInfo: PKInfo = decodeModel(jsonStr: message.content) else {
            self.messageManager.removeMessage(id: pkId)
            aui_info("acceptPK pkId: \(pkId) fail: pk info not found", tag: "PKService")
            completion?(NSError(domain: "pk info not found", code: 0, userInfo: nil))
            return
        }
        
        guard let roomPresenceInfo = roomPresenceService.getRoomPresenceInfo(roomId: pkInfo.fromRoomId),
              roomPresenceInfo.status == .idle else {
            self.messageManager.removeMessage(id: pkId)
            aui_info("acceptPK pkId: \(pkId) fail: room presence status is not idle", tag: "PKService")
            completion?(NSError(domain: "room presence status is not idle", code: 0, userInfo: nil))
            return
        }
        
        aui_info("acceptPK pkId: \(pkId)", tag: "PKService")
        pkInfo.type = .accept
        let content = encodeModelToJsonStr(pkInfo) ?? ""
        // 因为不管对方在不在线，更新presence总是会成功，这里被一个点对点消息用于判断对方是否在线
        messageManager.sendMessage(content: content, userId: pkInfo.fromUserId) {[weak self] err in
            guard let self = self else {return}
            aui_info("send accept message completion: \(err?.localizedDescription ?? "")", tag: "PKService")
            if let err = err {
                completion?(err)
                return
            }
            
            self.roomPresenceService.updateRoomPresenceInfo(roomId: self.channelName,
                                                            status: .pk,
                                                            interactorId: pkInfo.fromUserId,
                                                            interactorName: pkInfo.fromUserName) {[weak self] err in
                guard let self = self else { return }
                aui_info("acceptPK pkId: \(pkId) completion: \(err?.localizedDescription ?? "success")", tag: "PKService")
                if err == nil {
                    self.messageManager.removeMessage(id: message.id)
                }
                
                completion?(err)
            }
        }
        
    }
    
    public func rejectPK(pkId: String, completion: ((NSError?) -> ())?) {
        guard let message = messageManager.getMessage({ $0.content.contains(pkId) }),
              let pkInfo: PKInfo = decodeModel(jsonStr: message.content) else {
            aui_info("rejectPK pkId: \(pkId) fail: pk info not found", tag: "PKService")
            completion?(NSError(domain: "pk info not found", code: 0, userInfo: nil))
            return
        }
        aui_info("rejectPK pkId: \(pkId)", tag: "PKService")
        
        pkInfo.type = .reject
        let content = encodeModelToJsonStr(pkInfo) ?? ""
        messageManager.sendMessage(content: content,
                                   userId: pkInfo.userId) {[weak self] err in
            guard let self = self else { return }
            aui_info("rejectPK pkId: \(pkId) completion: \(err?.localizedDescription ?? "success")", tag: "PKService")
            if err == nil {
                self.messageManager.removeMessage(id: message.id)
            }
            
            completion?(err)
        }
    }
}


//MARK: InviteMessageProtocol
extension PKService: InviteMessageProtocol {
    public func onNewInviteDidReceived(channelName: String, message: InviteMessageInfo) {
        guard let info: PKInfo = decodeModel(jsonStr: message.content) else { return }
        aui_info("onNewInviteDidReceived message: \(message.content)", tag: "PKService")
        respDelegates.allObjects.forEach { delegate in
            delegate.onPKInfoDidReceive(channelName: channelName, info: info)
        }
    }
}

//MARK: RoomPresenceProtocol
extension PKService: RoomPresenceProtocol {
    public func onUserSnapshot(channelName: String, userList: [RoomPresenceInfo]) {
        
    }
    
    public func onUserUpdate(channelName: String, user: RoomPresenceInfo) {
        aui_info("onUserUpdate[\(user.roomId)] userId: \(user.ownerId), name: \(user.ownerName), status: \(user.status.rawValue)", tag: "PKService")
        let channelName = self.channelName
        // 被PK方 -> 发起PK方
        if user.roomId != channelName {
            if let currInfo = roomPresenceService.getRoomPresenceInfo(roomId: channelName),
               user.status == .pk,
               currInfo.status == .idle,
               user.interactorId == AUIRoomContext.shared.currentUserInfo.userId {
                aui_info("after acceptPK >> updateRoomPresenceInfo : status=\(InteractionType.pk.rawValue), interactorId=\(user.ownerId), interactorName=\(user.ownerName)", tag: "PKService")
                roomPresenceService.updateRoomPresenceInfo(roomId: channelName,
                                                           status: .pk,
                                                           interactorId: user.ownerId,
                                                           interactorName: user.ownerName,
                                                           completion: nil)
            }
        }
        
        // 发起PK方 -> 被PK方
        if user.roomId != channelName {
            if let currInfo = roomPresenceService.getRoomPresenceInfo(roomId: channelName),
               user.status == .pk,
               currInfo.status == .pk,
               currInfo.interactorId == user.ownerId,
               user.interactorId == currInfo.ownerId {
                aui_info("after acceptPK >> startPKInteraction roomId=\(user.roomId), ownerId=\(user.ownerId), ownerName=\(user.ownerName)", tag: "PKService")
                interactionService.startPKInteraction(roomId: user.roomId,
                                                      userId: user.ownerId,
                                                      userName: user.ownerName,
                                                      completion: nil)
                let info = self.createPKInfo(user, type: .accept)
                respDelegates.allObjects.forEach { delegate in
                    delegate.onPKInfoDidReceive(channelName: channelName, info: info)
                }
            }
        }
        
        // 发起PK方
        if user.roomId == channelName {
            if let pkRoomInfo = roomPresenceService.getRoomPresenceInfoByOwnerId(ownerId:  user.interactorId),
               user.status == .pk,
               pkRoomInfo.status == .pk,
               pkRoomInfo.interactorId == user.ownerId,
               user.interactorId == pkRoomInfo.ownerId {
                aui_info("after acceptPK >> startPKInteraction roomId=\(pkRoomInfo.roomId), ownerId=\(pkRoomInfo.ownerId), ownerName=\(pkRoomInfo.ownerName)", tag: "PKService")
                
                interactionService.startPKInteraction(roomId: pkRoomInfo.roomId,
                                                      userId: pkRoomInfo.ownerId,
                                                      userName: pkRoomInfo.ownerName,
                                                      completion: nil)
                let info = self.createPKInfo(pkRoomInfo, type: .accept)
                respDelegates.allObjects.forEach { delegate in
                    delegate.onPKInfoDidReceive(channelName: channelName, info: info)
                }
            }
        }
        
        
        // 被PK方/发起PK方 有一方关闭即停止PK
        if user.roomId != channelName {
            if let currInfo = roomPresenceService.getRoomPresenceInfo(roomId: channelName),
               user.status == .idle,
               currInfo.status == .pk,
               currInfo.interactorId == user.ownerId {
                aui_info("after stopPK >> stopInteraction", tag: "PKService")
                
                interactionService.stopInteraction { err in
                    aui_info("stopInteraction completion \(err?.localizedDescription ?? "success")")
                }
                
                let info = self.createPKInfo(currInfo, type: .end)
                info.fromUserId = user.ownerId
                info.fromUserName = user.ownerId
                info.fromRoomId = user.ownerName
                respDelegates.allObjects.forEach { delegate in
                    delegate.onPKInfoDidReceive(channelName: channelName, info: info)
                }
            }
        }
        
        // 主播已经和其他人PK
        if user.roomId != channelName {
            if let currInfo = roomPresenceService.getRoomPresenceInfo(roomId: channelName),
               user.status == .pk,
               currInfo.status == .pk,
               currInfo.interactorId == user.ownerId,
               user.interactorId != AUIRoomContext.shared.currentUserInfo.userId {
                aui_info("after startPK >> updateRoomPresenceInfo for pk owner has started interaction with others: status=\(InteractionType.idle.rawValue)", tag: "PKService")
                
                roomPresenceService.updateRoomPresenceInfo(roomId: channelName,
                                                           status: .idle,
                                                           interactorId: "",
                                                           interactorName: "",
                                                           completion: nil)
                
                let info = self.createPKInfo(currInfo, type: .end)
                respDelegates.allObjects.forEach { delegate in
                    delegate.onPKInfoDidReceive(channelName: channelName,info: info)
                }
                
            }
        }
        
    }
    
    public func onUserDelete(channelName: String, user: RoomPresenceInfo) {
        guard let info = interactionService.interactionInfo,
              info.type == .pk,
              info.userId == user.ownerId else { return }
        aui_info("onUserDelete ownerId: \(user.ownerId) ownerName: \(user.ownerName) roomId: \(user.roomId)", tag: "PKService")
        interactionService.stopInteraction(completion: nil)
        roomPresenceService.updateRoomPresenceInfo(roomId: channelName,
                                                   status: .idle,
                                                   interactorId: "", 
                                                   interactorName: "",
                                                   completion: nil)
    }
    
    public func onUserError(channelName: String, error: NSError) {
        
    }
}
